engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java

```diff
@@ -31,15 +39,78 @@ public final class RuntimeCache implements java.util.function.Function<String, O
    */
   @CompilerDirectives.TruffleBoundary
   public boolean offer(UUID key, Object value) {
+    expressions.put(key, new WeakReference<>(value));
     if (preferences.contains(key)) {
       var ref = new SoftReference<>(value);
       cache.put(key, ref);
-      expressions.put(key, new WeakReference<>(value));
+      Subject<Object> observable;
+      synchronized (observables) {
+        observable =
+            observables.computeIfAbsent(key, (UUID uuid) -> ReplaySubject.createWithSize(1));
+      }
+      observable.onNext(value);
       return true;
-    } else {
-      var ref = new WeakReference<>(value);
-      expressions.put(key, ref);
-      return false;
+    }
+    return false;
+  }
+
+  /**
+   * Registers a visualization for a given `expressionId`.
+   *
+   * @param visualizationId a unique ID of the visualization
+   * @param expressionId an expression's ID to which visualization is associated with
+   * @param onNext action to be executed with expression's value
+   * @param onFailure action to be executed when visualization throws a failure
+   * @param executor `Executor` on which new value notifications should be processed on
+   * @return true, if association of visualization with the expression was successful, false
+   *     otherwise
+   */
+  public boolean registerObserver(
+      UUID visualizationId,
+      UUID expressionId,
+      Consumer<Object> onNext,
+      Consumer<Throwable> onFailure,
+      Executor executor) {
+    Subject<Object> observable;
+    synchronized (observables) {
+      observable =
+          observables.computeIfAbsent(expressionId, (UUID uuid) -> ReplaySubject.createWithSize(1));
+    }
+    var scheduler = Schedulers.from(executor);
+    var observer =
```

Comment:
- am I wrong to say that this is just like `CompletableFuture` that others may "chain" actions to
- which is triggered when the value is changed?
- for something as simple we probably don't need `rxjava` 3rd party library
- there are methods like `thenApply(..., executor)` already in the JDK...
- but when talking about _"reactive cache"_ I probably wanted something completely different